extensions/amp-story/1.0/amp-story-tooltip.js

```diff
@@ -0,0 +1,401 @@
+            this.tooltipOverlayEl_.setAttribute('desktop', '') :
+            this.tooltipOverlayEl_.removeAttribute('desktop');
+        });
+  }
+
+  /**
+   * Builds tooltip and attaches it depending on the clicked element content and
+   * position.
+   * @param {!Element} clickedEl
+   * @private
+   */
+  attachTooltipToEl_(clickedEl) {
+    const {href} = parseUrlDeprecated(clickedEl.getAttribute('href'));
+
+    const domainName = getSourceOriginForElement(clickedEl, href);
+    const tooltipText = clickedEl.getAttribute('tooltipText') || domainName;
+    this.appendTextToTooltip_(tooltipText);
+
+    const iconAttr = clickedEl.getAttribute('tooltipIcon');
+    const iconSrc = iconAttr ? parseUrlDeprecated(iconAttr).href : '';
+    this.appendIconToTooltip_(iconSrc);
+
+    this.positionTooltip_(clickedEl);
+  }
+
+  /**
+   * Checks for existing text content and modifies it or appends a new text node
+   * if building tooltip for the first time.
+   * @param {string} tooltipText
+   * @private
+   */
+  appendTextToTooltip_(tooltipText) {
+    const existingTooltipText = childElement(
+        dev().assertElement(this.tooltip_),
+        el => el.classList.contains('i-amphtml-tooltip-text'));
+
+    if (existingTooltipText) {
+      existingTooltipText.textContent = tooltipText;
+    } else {
+      const html = htmlFor(this.win_.document);
+      const tooltipTextEl =
+        html`<p class="i-amphtml-tooltip-text" ref="text"></p>`;
+      tooltipTextEl.textContent = tooltipText;
+      this.tooltip_.insertBefore(tooltipTextEl, this.tooltip_.firstChild);
+    }
+  }
+
+  /**
+   * Checks for existing icon source and modifies it or appends a new tooltip
+   * icon node if building tooltip for the first time.
+   * @param {string} iconSrc
+   * @private
+   */
+  appendIconToTooltip_(iconSrc) {
+    const existingTooltipIcon = childElement(
+        dev().assertElement(this.tooltip_),
+        el => el.classList.contains('i-amphtml-story-tooltip-icon'));
+
+    if (!iconSrc) {
+      if (existingTooltipIcon) {
+        removeElement(existingTooltipIcon);
+      }
+      return;
+    }
+
+    if (existingTooltipIcon && existingTooltipIcon.firstElementChild) {
+      addAttributesToElement(existingTooltipIcon.firstElementChild,
+          dict({'src': iconSrc}));
+    } else {
+      const html = htmlFor(this.win_.document);
+      const iconEl =
+        html`
+        <div class="i-amphtml-story-tooltip-icon">
+          <img ref="icon">
+          </img>
+        </div>`;
+      const {icon} = htmlRefs(iconEl);
+      addAttributesToElement(icon, dict({'src': iconSrc}));
+      this.tooltip_.insertBefore(iconEl, this.tooltip_.firstChild);
+    }
+  }
+
+  /**
+   * Positions tooltip and its pointing arrow according to the position of the
+   * clicked element.
+   * @param {!Element} clickedEl
+   * @private
+   */
+  positionTooltip_(clickedEl) {
+    let clickedRect, storyWidth, storyHeight, tooltipWidth, tooltipTop,
```

Comment:
Please differentiate what's used in the `measure`, and what needs to be passed from the `measure` to the `mutate`
For what's in the measure only, you can simply use `const` in the measure callback.
For what needs to be passed, please declare a `state` object that you fill from the measure, and use in the mutate.

`const state = {};`